packages/block-editor/src/components/date-format-picker/index.js

```diff
@@ -0,0 +1,170 @@
+/**
+ * External dependencies
+ */
+import { uniq } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { _x, __ } from '@wordpress/i18n';
+import { dateI18n } from '@wordpress/date';
+import { useState, createInterpolateElement } from '@wordpress/element';
+import {
+	TextControl,
+	ExternalLink,
+	VisuallyHidden,
+	__experimentalToggleGroupControl as ToggleGroupControl,
+	__experimentalToggleGroupControlOption as ToggleGroupControlOption,
+	CustomSelectControl,
+	BaseControl,
+} from '@wordpress/components';
+
+// So that we can illustrate the different formats in the dropdown properly,
+// show a date that has a day greater than 12 and a month with more than three
+// letters. Here we're using 2022-01-25 which is when WordPress 5.9 was
+// released.
+const EXAMPLE_DATE = new Date( 2022, 0, 25 );
+
+/**
+ * The `DateFormatPicker` component renders controls that let the user choose a
+ * _date format_. That is, how they want their dates to be formatted.
+ *
+ * @see https://github.com/WordPress/gutenberg/blob/HEAD/packages/block-editor/src/components/date-format-picker/README.md
+ *
+ * @param {Object}                          props
+ * @param {string|null}                     props.format        The selected date
+ *                                                              format.
+ * @param {string}                          props.defaultFormat The date format that
+ *                                                              will be used if the
+ *                                                              user selects
+ *                                                              'Default'.
+ * @param {( format: string|null ) => void} props.onChange      Called when a
+ *                                                              selection is made.
+ */
+export default function DateFormatPicker( {
+	format,
+	defaultFormat,
+	onChange,
+} ) {
+	return (
+		<fieldset className="block-editor-date-format-picker">
+			<VisuallyHidden as="legend">{ __( 'Date format' ) }</VisuallyHidden>
+			<div className="block-editor-date-format-picker__header">
+				{ __( 'Format' ) }
+				<span className="block-editor-date-format-picker__header__hint">
+					{ dateI18n( format || defaultFormat, EXAMPLE_DATE ) }
+				</span>
+			</div>
+			<ToggleGroupControl
+				label={ __( 'Date format' ) }
+				isBlock
+				hideLabelFromVision
+				value={ format ? 'custom' : 'default' }
+				onChange={ ( value ) => {
+					if ( value === 'default' ) {
+						onChange( null );
```

Comment:
It's interesting that it gets set to `null`. That possibly presents the implementor of the component with a gotcha, they have to ensure they implement a fallback to the `defaultFormat` themselves.

I feel like it might be easier to implement if `format` acted more like the `value` of a controlled input and `onChange` always returned the selected format.

An option could be to not show the default format in the `CustomSelectControl` as an option, then the logic within the component becomes a bit simpler, and I think it also removes some overhead for both the implementor and the end user (because I don't they need two different ways to select it).

It is a design decision though.

Reply:
Maybe I am not understanding you properly (it's getting late on a Friday lol) but I think this is right because what you're indicating when you choose _Default_ is that _I don't have a preference_. It's similar to the layout controls where one of the options is to choose _Inherit default layout_. We're using `null` to indicate _nothing_ which is semantically correct.

Reply:
To put it another way I think there's a meaningful difference between selecting _Default_ and getting `M i, Y` and selecting _Custom_ and choosing `M i, Y`. The latter wouldn't change if you update the setting in Options → General.

Reply:
Ah, yes you're right, that wouldn't work if the component always returns the default as its value. 👍 

It's a shame it requires using `null` to signal that the default should be used, but I can't think of a better way. You've probably thought about this a lot too.

Maybe a little bit of extra detail could be added to the doc block for `props.format` to document that the `null` value means the default has been chosen.

Reply:
👍 good idea ce9835d152